zarr/storage.py

```diff
@@ -1877,6 +1877,141 @@ def __delitem__(self, key):
             self._invalidate_value(key)
 
 
+class SQLiteStore(MutableMapping):
+    """Storage class using SQLite.
+
+    Parameters
+    ----------
+    path : string
+        Location of database file.
+    **kwargs
+        Keyword arguments passed through to the `sqlite3.connect` function.
+
+    Examples
+    --------
+    Store a single array::
+
+        >>> import zarr
+        >>> store = zarr.SQLiteStore('data/array.sqldb')
+        >>> z = zarr.zeros((10, 10), chunks=(5, 5), store=store, overwrite=True)
+        >>> z[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+
+    Store a group::
+
+        >>> store = zarr.SQLiteStore('data/group.sqldb')
+        >>> root = zarr.group(store=store, overwrite=True)
+        >>> foo = root.create_group('foo')
+        >>> bar = foo.zeros('bar', shape=(10, 10), chunks=(5, 5))
+        >>> bar[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+    """
+
+    def __init__(self, path, **kwargs):
+        import sqlite3
+
+        kwargs.setdefault('timeout', 5.0)
+        kwargs.setdefault('detect_types', 0)
+        kwargs.setdefault('isolation_level', None)  # autocommit
+        kwargs.setdefault('check_same_thread', False)  # disallow writing from other threads
```

Comment:
From the [docs](https://docs.python.org/3/library/sqlite3.html#sqlite3.connect):

> By default, check_same_thread is True and only the creating thread may use the connection. If set False, the returned connection may be shared across multiple threads. When using multiple threads with the same connection writing operations should be serialized by the user to avoid data corruption.

Maybe we should set this to True. The docs say any multithreaded use needs to be fully synchronized by the user. There is no mechanism in zarr for doing this - the synchronizer classes only prevent writing the same chunk at the same time, but they allow different chunks to be written concurrently, which the sqlite docs suggest could still cause corruption.

Reply:
Yep, that was probably an oversight on my part. Have corrected this to `True`.

I think if we just make sure to `flush` after every write operation, we should be safe allowing this though.

Reply:
After giving this some thought, it seems reasonable for us to always `flush` after mutating the database. For example other processes could be accessing the same database and we want to ensure they all see the same thing. This isn't too different from how we handle other stores.

Have pushed a few commits to tidy things up and implement this `flush`ing with each mutation behavior. Also have added an `update` function to allow more efficient submission of multiple changes (so we only commit once).

Switched back to setting `check_same_thread` to `False` by default as we now always serialize data if a change is made.